Fix #21417 - Cannot view zoomed in high up

MaxMallon opened the pull request:
For [21417](https://github.com/OpenRCT2/OpenRCT2/issues/21417)

Gymnasiast commented on src/openrct2/world/Map.h:
```diff
@@ -37,6 +37,7 @@ constexpr uint8_t ConstructionRightsClearanceSmall = 3;
 constexpr const uint8_t ConstructionRightsClearanceBig = 3 * COORDS_Z_STEP;
 
 #define MAP_MINIMUM_X_Y (-MAXIMUM_MAP_SIZE_TECHNICAL)
+#define MAP_Z_PARALAX_OFFSET (1800)
```
Please create a constexpr, not a define - those are remnants from our C days.

Gymnasiast commented:
This is going to need extensive testing - the potential for unforeseen edge cases is quite big.

MaxMallon commented:
> This is going to need extensive testing - the potential for unforeseen edge cases is quite big.

Anything I can test to help speed it up?

733737 commented:
This will be a great boon to my buildstyle. I could provide test parks if you'd like

deurklink commented:
This is quite often an issue in our mini contests where I want to restrict people to building on a small map, so people want to go more vertical

Gymnasiast commented on src/openrct2-ui/windows/TopToolbar.cpp:
```diff
@@ -1700,7 +1700,7 @@ class TopToolbar final : public Window
         auto screenPos = sourceScreenPos;
         uint16_t maxPossibleHeight = ZoomLevel::max().ApplyTo(
-            std::numeric_limits<decltype(TileElement::BaseHeight)>::max() - 32);
+            std::numeric_limits<decltype(TileElement::BaseHeight)>::max() - 1);
```
This is unrelated to the fix for stuff getting cut off. In fact, this makes the problems more likely to occur.

MaxMallon replied:
> This is unrelated to the fix for stuff getting cut off. In fact, this makes the problems more likely to occur.

Yeah, that's what I mentioned in the discord, I'm not sure why this even showed up in here. That's from a completely different PR, where I fixed being able to place walls and small scenery higher up.
What about it makes problems more likely, though? I've also been using that change locally, and it's worked wonderfully.

Gymnasiast commented:
Now I’ve had time to look at it, I’m seriously puzzled at a lot of things.

First and foremost that you expect us to review and merge something like this in 3 days. This is almost impossible in the best of times, but this is not the best of times. Your branch contains unrelated commits, the commit messages do not adhere to the rules, there was no issue, no affected park or any screenshots, you just told us "here is a problem, please merge this immediately". We really can’t work like this.

For stuff like this, an issue is needed. _You_ may think this is the best solution, and you may certainly suggest it in an issue, but we prefer deciding it based on the problem at hand. We recently had to close a few PRs due to this problem.

Also, I find this PR in relation to #21368 to be a little bit suspect. That ceiling was there for a good reason, and I have the sneaking suspicion that hacking around it (e.g. with the Tile Inspector) could very well be the reason for the problem this PR purports to fix. I have no way of telling, though, as you didn’t create an issue or provided the necessary information or test parks, so I’m essentially left guessing.

MaxMallon commented:
> For stuff like this, an issue is needed. 
You yourself said an issue wasn't needed when I asked in the discord... 

I stumbled upon a problem while playing the game, and fixed it. It's not like this is an elaborate new feature, where discussion is needed. It's just a fix, for something that's very clearly a problem. 

> That ceiling was there for a good reason
What was the reason?

>I have the sneaking suspicion that hacking around it (e.g. with the Tile Inspector) could very well be the reason for the problem this PR purports to fix. 
Well no, you can easily make a park without any cheats on, without tile inspector, that runs into the same issue. Terrain itself can be raised to the build height, and you could even build scenery on top of that terrain, above the height limit used when shifting, resulting in the situation where you could build on top of the land, but not at the same height next to it.

Gymnasiast commented:
> You yourself said an issue wasn't needed when I asked in the discord...
>
> I stumbled upon a problem while playing the game, and fixed it. It's not like this is an elaborate new feature, where discussion is needed. It's just a fix, for something that's very clearly a problem.

I’m not here to argue. Please add the requested things.

Gymnasiast reviewed (commented):
I rebased the PR and removed the unrelated stuff. I also fixed the spelling of “parallax”, which you hadn’t fixed despite me asking.

I have now tested it. As I was fearing, it makes the void on parks like Micro Park _enormous_ - it just keeps on scrolling, making it almost impossible to find it back without zooming out or using the Map window.

It’s not difficult to see why this is the case, as you simply add a big fixed margin to it. Rather, it should take the actual height into account - which would also work better for your case.

MaxMallon commented:
> I rebased the PR and removed the unrelated stuff. I also fixed the spelling of “parallax”, which you hadn’t fixed despite me asking.

I was already on it myself, but thanks. Also you never pointed out the misspelling of parallax. Ctrl+f-fing only shows you saying such now, probably now me saying it here. 

> I have now tested it. As I was fearing, it makes the void on parks like Micro Park _enormous_ - it just keeps on scrolling, making it almost impossible to find it back without zooming out or using the Map window.

I was afraid of that yeah, but I figured it would be less of an issue than the current situation. 

> It’s not difficult to see why this is the case, as you simply add a big fixed margin to it. Rather, it should take the actual height into account - which would also work better for your case.

That would be ideal, and I had considered it. The question is, what height? The function that converts the 3d to 2d coordinates takes height 0 now. But detecting how the thing you're currently looking at would be, well, not easy. Do we add a park-wide 'tallness' variable, that just takes the highest thing in the park?

733737 commented:
I think that you should implement a method of detecting whether or not the park is tall and only then modify the viewport

AaronVanGeffen commented:
> Also you never pointed out the misspelling of parallax.

It was pointed out in a line comment: https://github.com/OpenRCT2/OpenRCT2/pull/21404#discussion_r1489700537

AaronVanGeffen commented:
> I think that you should implement a method of detecting whether or not the park is tall and only then modify the viewport

Sure, but then when is a park tall? Check the height of the outer X tiles on either side? But then what if the X+1th tile is just very high? It quickly becomes computationally expensive to query this information this way.

We'll need a good heuristic rather than something that is applied in all situations, I agree.

MaxMallon commented:
> > Also you never pointed out the misspelling of parallax.
> 
> It was pointed out in a line comment: [#21404 (comment)](https://github.com/OpenRCT2/OpenRCT2/pull/21404#discussion_r1489700537)

I .. still don't see it.. but I'll take your word for it then.

ZehMatt commented:
I don't think the solution this PR offers it the correct approach for this specific problem, ideally this will take the zoom level into the equation so that when the user zooms back in from far away he won't be lost in the void. Also another problem I have is with this magic constant, what is the meaning of the value and how did you get there? Does this value cover other situations? The hardcoded value leads to more questions than answers currently, when taking the zoom level into the equation this should be probably calculated so that on lower zoom levels its still close enough to the map borders. I understand that there is a need for this but this doesn't seem like the proper way to do it.

MaxMallon commented:
> I .. still don't see it.. but I'll take your word for it then.

Now I see it, but I swear that wasn't there when I read that comment for the 1st time, else I would've changed it ofc.

MaxMallon commented:
> I don't think the solution this PR offers it the correct approach for this specific problem, ideally this will take the zoom level into the equation so that when the user zooms back in from far away he won't be lost in the void. Also another problem I have is with this magic constant, what is the meaning of the value and how did you get there? Does this value cover other situations? The hardcoded value leads to more questions than answers currently, when taking the zoom level into the equation this should be probably calculated so that on lower zoom levels its still close enough to the map borders. I understand that there is a need for this but this doesn't seem like the proper way to do it.

To get the 1800 I placed 4 entities at z=2000, which is about the highest you can get them, in the corners of a map, and looked at how big the offset needed to be roughly, to still keep them centered at maximum zoom. About 1800. Actually, the zoom doesn't matter, whether you're zoomed in fully, or fully zoomed out, the view still can't center without that extra offset. The only difference is that if you're zoomed out, you have a bigger view, and so while the entertainer is not centered, you still see him.

ZehMatt commented:
> > I don't think the solution this PR offers it the correct approach for this specific problem, ideally this will take the zoom level into the equation so that when the user zooms back in from far away he won't be lost in the void. Also another problem I have is with this magic constant, what is the meaning of the value and how did you get there? Does this value cover other situations? The hardcoded value leads to more questions than answers currently, when taking the zoom level into the equation this should be probably calculated so that on lower zoom levels its still close enough to the map borders. I understand that there is a need for this but this doesn't seem like the proper way to do it.
> 
> To get the 1800 I placed 4 entities at z=2000, which is about the highest you can get them, in the corners of a map, and looked at how big the offset needed to be roughly, to still keep them centered at maximum zoom. About 1800. Actually, the zoom doesn't matter, whether you're zoomed in fully, or fully zoomed out, the view still can't center without that extra offset. The only difference is that if you're zoomed out, you have a bigger view, and so while the entertainer is not centered, you still see him.

I will have a closer look at the PR and do some testing once I finish some of the other things I'm working on at the moment, I worked on something like this a few years back and it introduced a bunch of problems.

MaxMallon commented:
> I worked on something like this a few years back and it introduced a bunch of problems.

What sort of problems?

ocalhoun6 commented:
> Sure, but then when is a park tall? Check the height of the outer X tiles on either side? But then what if the X+1th tile is just very high? It quickly becomes computationally expensive to query this information this way.

First, have a park-wide variable for 'tall park', with default to false.

Then we need to go into *every* possible function for placing every possible object -- path, scenery, surface, rides, etc, etc, etc. Inside each of those functions, place a check to set the new variable: If (the height you're placing this object at) > (magic number), set 'tall park' = true.

Pros:
- You only need one fairly inexpensive check for each time an object is placed. Shouldn't impact performance too much. (No need to iterate over every single tile in the park, no need to run it every tick or anything.)
- It will accomplish the goal of only widening the viewable area if the park is tall.

Cons:
- It won't affect pre-existing tall parks unless you add a new object of some kind to them. (This could also be overcome by adding a check on import, but that *would* require iterating over all the tiles. Still, if it's only done once, when loading the save, it might not be *that* big of a deal.)
- If you build very tall stuff in a park and then remove all the tall stuff, the park will *not* revert to the 'tall park = false' view mode. (This could be hand-waved away by saying, "Well, if you built very tall things in a park once, you likely want to do so again, so the view makes sure this is still possible.) (If the check on import is implemented, then this con could also be ameliorated: save the park, close it, and reopen the file, and the view will return to normal if there's no longer any tall objects in the park. Perhaps a cheat or console command could also be added, allowing the user to manually set or unset the 'tall park' variable.)
- Requires a check on every possible method to move or place every possible object ... it would be easy to miss something and thereby introduce a bug.

Overall, I think that's a pretty good way to implement this without being too computationally expensive.

ZehMatt commented:
After some testing it is pretty much to what I expected, the view is allowed to go way too far away from the map, the map window also doesn't quite like this, the cap is still there when navigating with the cursor but with the arrow keys you can further so this would need to be changed as well. I know that this PR will allow the view to work with some high up build things but the way its trying to solve it is rather sloppy, if it were to restrict the view to the visible area of the map then I would be absolutely fine with such PR but this has too many side effects so I can't really approve this.

MaxMallon commented:
> the map window also doesn't quite like this

To be fair, the map has been pretty damn useless for ages now, it's always surrounded with endless void, the actual park is not centered in the scroll bars, and very large parks don't fit on it anyway.

Look, I'm not saying my solution is perfect, it's very clearly not. But it _is_ better than the current situation. I also tried micropark, and while it's true that you can scroll ludicrously far into the void with this fix, .. why would you? It's a micropark, you don't need to scroll at all. Even if you somehow do it by accident, yanking the mouse or something, it's not exactly hard to find back the center. Whereas in the current situation, if you want to view high up stuff like this.. you flat-out _cannot._ 
So, current situation:
Low park = fine
High park = impossible

fix situation:
Low park = fine but with minor worsening due to too much scroll
High park = fine

I'd say that's a net-gain. 
And after we can of course still keep looking for a better solution after. 

Also, 
>the cap is still there when navigating with the cursor but with the arrow keys you can further so this would need to be changed as well.

I'm not sure what you mean with this, I tried both scrolling with the cursor, which is what I always use myself, and with the arrows (which I forgot was a thing at all), and both worked fine?

ZehMatt commented:
I know this approach isn't perfect but it makes a couple of things worse which no PR should do, building things that high is not how people play the game most of the time also. I would also appreciate if you would stop arguing and rather find a better solution if you want this to be changed.

733737 commented:
If micropark is the example that doesn't work, why not use an equation that multiplies by the x and y dimensions before adding to the scroll distance

MaxMallon commented:
> If micropark is the example that doesn't work, why not use an equation that multiplies by the x and y dimensions before adding to the scroll distance

because it's a fixed offset, the height. whether your park is 3x3, or 999x999, the extra 'distance' needed to go into the void, to see stuff at max height is the same.

> but it makes a couple of things worse which no PR should do,

I mean, the PR that introduced the new safe file format broke backwards compatibility, which is also making a thing worse; yet no-one (me included) would argue against that as the benefits easily outweigh the downsides. 

>  building things that high is not how people play the game most of the time 

Probably because they can't view up there. and even then, multiple people indicate a desire for this to be fixed.

>  I would also appreciate if you would stop arguing

Fine, then this will be the last thing. Since I don't know a better solution, guess this simply won't get fixed. Guess this is another thing to add to the list of hours spent on PRs that you guys reject. Why do I even bother anymore

733737 commented:
I think that in micropark, there is less need for absolute maximum height visibility, so you could just do a percentage of maximum height increase improvement. This is a compromise from the all or nothing increase.

Gymnasiast commented:
> Fine, then this will be the last thing. Since I don't know a better solution, guess this simply won't get fixed. Guess this is another thing to add to the list of hours spent on PRs that you guys reject. Why do I even bother anymore

You brought that upon yourself, I’m afraid. You refused to discuss any of it upfront and then demanded we merged it before the contest deadline. That is not a recipe for success, to say the least.
